packages/core/src/tracing/clsHooked/async_local_storage_context.js

```diff
@@ -0,0 +1,361 @@
+              // this.exit(context);
+              return result;
+            })
+            .catch((/** @type {*} */ err) => {
+              // this.exit(context);
+              throw err;
+            });
+        }
+      });
+    } finally {
+      if (!isPromise) {
+        // fn did not return a promise, so we behave like this.runAndReturn.
+        // this.exit(context);
+      }
+    }
+  }
+
+  /**
+   * Returns a wrapper around the given function which will enter CLS context which is active at the time of calling
+   * bind and leave that context once the function terminates. If no context is active, a new context will be created.
+   * @param {Function} fn
+   * @param {InstanaCLSContext} context
+   */
+  bind(fn, context) {
+    if (!context) {
+      const activeContext = storage.getStore();
+      if (!activeContext) {
+        context = this.createContext();
+      } else {
+        context = activeContext;
+      }
+    }
+
+    // const self = this;
+
+    return function clsBind() {
+      storage.enterWith(context);
+      try {
+        return fn.apply(this, arguments);
+      } finally {
+        // self.exit(context);
+      }
+    };
+  }
+
+  /**
+   * Binds the given emitter to the currently active CLS context. Work triggered by an emit from that emitter will
+   * happen in that CLS context.
+   * @param {import('events').EventEmitter} emitter
+   */
+  bindEmitter(emitter) {
+    assert.ok(emitter.on && emitter.addListener && emitter.emit, 'can only bind real EEs');
+
+    const namespace = this;
+    const thisSymbol = `context@${this.name}`;
+
+    /**
+     * Capture the context active at the time the emitter is bound.
+     * @param {*} listener
+     */
+    function attach(listener) {
+      if (!listener) {
+        return;
+      }
+      if (!listener[CONTEXTS_SYMBOL]) {
+        listener[CONTEXTS_SYMBOL] = Object.create(null);
+      }
+
+      listener[CONTEXTS_SYMBOL][thisSymbol] = {
+        namespace,
+        context: storage.getStore()
+      };
+    }
+
+    /**
+     * At emit time, bind the listener within the correct context.
+     * @param {*} unwrapped
+     */
+    function bind(unwrapped) {
+      if (!(unwrapped && unwrapped[CONTEXTS_SYMBOL])) {
+        return unwrapped;
+      }
+
+      let wrapped = unwrapped;
+      const unwrappedContexts = unwrapped[CONTEXTS_SYMBOL];
+      Object.keys(unwrappedContexts).forEach(name => {
+        const thunk = unwrappedContexts[name];
+        wrapped = thunk.namespace.bind(wrapped, thunk.context);
+      });
+      return wrapped;
+    }
+
+    wrapEmitter(emitter, attach, bind);
+  }
+
+  /**
+   * @param {InstanaCLSContext} context
+   */
+  enter(context) {
+    assert.ok(context, 'context must be provided for entering');
+    storage.enterWith(context);
+    this.sharedContext = context;
+  }
+
+  /**
+   * @param {InstanaCLSContext} context
+   */
+  exit(context) {
+    assert.ok(context, 'context must be provided for exiting');
```

Comment:
Should we set this.sharedContext = null here?

Reply:
Nevermind, that would probably defeat the purpose of having it in the first place.